Restructure of Car Vehicle for external firmware control

rajat2004 opened the pull request:
This PR restructures the Car vehicle code to make it similar to Multirotor in terms of adding support for external firmware control of the car vehicle
An example of how to add external firmware control can be seen here - https://github.com/microsoft/AirSim/commit/a5dce95d82506741412f26072e3d4b7e6f20cdd9 (ArduPilot's Rover vehicle, the entire branch is https://github.com/microsoft/AirSim/compare/master...rajat2004:pr-ardurover2)
(Didn't open PR for the entire thing since would be more difficult to review)

This PR has been tested to be working with manual control from the keyboard on Ubuntu 16.04 and Windows 10, with the below Python APIs on Ubuntu *(Unreal Engine)*

Things tested to be working - 
* Movement APIs - `hello_car.py`
* getCarState - `car_monitor.py`
* Pause - `pause_continue_car.py`
* Multi-vehicle - `multi_agent_car.py`
* ClockSpeed

Unity - Compilation only till now, real testing soon to follow

There are a few more things which can be cleaned up more, such as removing the `CarState` and `CarControls` from `CarApiBase`
Feedback on the approach by the other devs is very much needed!

rajat2004 commented:
Review requested @madratman @sytelus

sytelus commented:
@madratman - this looks good to me. Please feel free to merge if you are happy as well!

rajat2004 committed:
Restructure of Car Vehicle for external firmware control

Split CarApiBase implementation from CarPawnApi and add CarApiFactory to create appropriate vehicle API based on settings

rajat2004 commented:
Rebased on master

rajat2004 commented:
Will need to make changes to the Unity code also, so would be great to have a review!
I'll start on the Unity ones, mostly shouldn't be much different

rajat2004 committed:
Unity Car Restructure for external firmware control

rajat2004 commented:
Preliminary Unity changes added, tested till compilation only (Needs https://github.com/microsoft/AirSim/pull/2200)
Will test it on Unity as well in sometime

Jaeyoung-Lim commented:
Pretty cool! Any plans on getting this merged?

madratman commented:
Sorry, been busy with other things. Reviewing and testing rn.

ironclownfish commented on AirLib/include/vehicles/car/api/CarApiBase.hpp:
```diff
@@ -127,16 +131,16 @@ class CarApiBase : public VehicleApiBase  {
     SensorCollection sensors_; //maintains sensor type indexed collection of sensors
     vector<unique_ptr<SensorBase>> sensor_storage_; //RAII for created sensors
 
-protected:
```
Why is resetImplementation() no longer protected? Other classes should always call reset()--which calls resetImplementation()--so I'm wary of resetImplementation() being public.

ironclownfish commented on AirLib/include/vehicles/car/api/CarApiBase.hpp:
```diff
@@ -127,16 +131,16 @@ class CarApiBase : public VehicleApiBase  {
     SensorCollection sensors_; //maintains sensor type indexed collection of sensors
     vector<unique_ptr<SensorBase>> sensor_storage_; //RAII for created sensors
 
-protected:
```
resetImplementation() should be protected. Other classes should call reset(), which internally checks for errors and then calls resetImplementation().

rajat2004 replied:
Changes made

ironclownfish commented on AirLib/include/vehicles/car/firmwares/physxcar/PhysXCarApi.hpp:
```diff
@@ -0,0 +1,97 @@
+    {}
+
+public:
+    virtual void resetImplementation() override
```
Should be protected

rajat2004 replied:
Changed

ironclownfish commented on Unity/AirLibWrapper/AirsimWrapper/Source/Vehicles/Car/CarPawnApi.h:
```diff
@@ -4,32 +4,27 @@
+    void updateMovement(const msr::airlib::CarApiBase::CarControls& controls);
+    msr::airlib::CarApiBase::CarState getCarState() const;
+
+    void resetImplementation();
```
make protected

ironclownfish commented on Unity/AirLibWrapper/AirsimWrapper/Source/Vehicles/Car/CarPawnSimApi.cpp:
```diff
@@ -169,13 +171,13 @@ void CarPawnSimApi::resetImplementation()
 
 	PawnSimApi::resetImplementation();
-	vehicle_api_->reset();
+	phys_api_->resetImplementation();
```
call phys_api_->reset(), which calls resetImplementation() internally.

rajat2004 replied:
I've renamed the function to `reset`, but not sure if that's the correct fix
It's calling ``PhysXApiBase's reset()

ironclownfish commented on Unity/AirLibWrapper/AirsimWrapper/Source/Vehicles/Car/CarPawnSimApi.h:
```diff
@@ -19,7 +20,7 @@ class CarPawnSimApi : public PawnSimApi
-	CarPawnSimApi(const Params& params, const CarPawnApi::CarControls&  keyboard_controls, std::string car_name);
+    CarPawnSimApi(const Params& params, const msr::airlib::CarApiBase::CarControls& keyboard_controls, std::string car_name);
 	virtual ~CarPawnSimApi() = default;
 	virtual void resetImplementation() override;
```
protected

rajat2004 replied:
Done

bmalhigh commented:
@rajat2004 -- Good work !! Couple of clarifying questions for my understanding --

1) Is it correct that before this change, carcontrols coming via API were directly being applied to Unreal UWheeledVehicleMovementComponent without waiting for tick/update and now with your changes, the controls coming via API are also applied only during tick/update. correct? I don't see any issue with that but just wanted to understand the change clearly.
2) Do we still need CarPawnApi class object or can that work be done by CarPawn class? The way you have done it now is definitely more evolving/less-disrupting... but trying to understand if CarPawnAPI still serves a very distinct role or not (or if this is going to more confusion w.r.t. number of objects/types)

rajat2004 commented:
Sorry for the delayed response. was travelling. And thanks for the review!

@ironclownfish I have added a separate commit - https://github.com/microsoft/AirSim/pull/2172/commits/74518e36ebd0b8aece8b406fd6dc3ce47361f0a1 for the changes since I'm not very sure if I've done it in the correct way, once any additional changes needed are done, will squash it

> Is it correct that before this change, carcontrols coming via API were directly being applied to Unreal UWheeledVehicleMovementComponent without waiting for tick/update and now with your changes, the controls coming via API are also applied only during tick/update. correct? I don't see any issue with that but just wanted to understand the change clearly.

@bmalhigh Not exactly, the behaviour is the same as in the previous code, the main thing which I've done is the removal of `CarPawnApi` as the child of `CarApiBase` and added a seperate firmware API interface (`PhysXCarApi`) with which the physics would interface.
Earlier, the CarControls coming via API were being applied only on RenderTicks (rather than on Physics Ticks - explained by @sytelus in https://github.com/microsoft/AirSim/issues/2102#issuecomment-529184897)
It's the same with these changes other than that the controls are fetched from the firmware API interface and then applied to the MovememntComponent

> Do we still need CarPawnApi class object or can that work be done by CarPawn class? The way you have done it now is definitely more evolving/less-disrupting... but trying to understand if CarPawnAPI still serves a very distinct role or not (or if this is going to more confusion w.r.t. number of objects/types)

This PR does reduce the work done by CarPawnApi class quite a lot, but I feel it's cleaner to keep it seperate from CarPawn class, gives a easy-to-read interface 
(Incidently, I haven't even looked at the CarPawn class much, quite a lot of Unreal code in that)
Just my opinion, I'll leave it upto the maintainers to decide whether to keep it or not

Hope this explains it a bit clearly

rajat2004 commented:
I've added some Unity changes also, will be squashed after everything is okay

rajat2004 commented:
Been quite some time since I updated this, life got in between, with exams, sickness, system replacement, etc..
Will try to get this done as soon as possible, Unity wasn't working last I tried, would be great to have some pointers

rajat2004 commented:
I think this is ready to be merged now, both Unreal and Unity are working, commits have been squashed

rajat2004 commented:
Ping!

rajat2004 commented:
Opened https://github.com/microsoft/AirSim/pull/2383 which has the additional commits for ArduRover support

madratman commented on Unreal/Plugins/AirSim/Source/Vehicles/Car/CarPawnSimApi.cpp:
```diff
@@ -20,16 +20,17 @@ void CarPawnSimApi::initialize()
+
+    vehicle_api_ = CarApiFactory::createApi(getVehicleSetting(), sensor_factory, (*getGroundTruthKinematics()), 
+                                            (*getGroundTruthEnvironment()), home_geopoint);
+    phys_api_ = std::unique_ptr<CarPawnApi>(new CarPawnApi(pawn, getGroundTruthKinematics(), vehicle_api_.get()));
```
PR looks good to me. 
In general, AirSim does have a few inconsistencies wrt multirotor and car in apis / update thread. 
- pointers v/s references for kinematics, env
- multirotor: all data exchange taking inside `u/pdateRenderedState` and unreal pawn pose update happens in `updateRendering`, whereas in car: both happen in `updateRendering`

rajat2004 replied:
Will try to reduce some of the inconsistencies between them, relating to the pointers vs references
The `updateRendering` ones will probably a much bigger task and a scope-creep in this PR, can be done in a follow-up one

madratman replied:
Yeah I think the ptr vs ref is also not your fault. It's there historically. The two points I mentioned do need refactoring.

madratman commented on Unity/AirLibWrapper/AirsimWrapper/Source/Vehicles/Car/CarPawnSimApi.h:
```diff
@@ -33,15 +34,19 @@ class CarPawnSimApi : public PawnSimApi
     Params params_;
 
 	std::unique_ptr<msr::airlib::CarApiBase> vehicle_api_;
+	std::unique_ptr<CarPawnApi> phys_api_;
```
I think - calling this `phys_api_` is a bit misleading? It's akin to `AFlyingPawn` BP, but in C++ right? 
Perhaps `unreal_pawn_api_` or simply `pawn_api_` is more suitable.

madratman replied:
I agree with your old reply - `This PR does reduce the work done by CarPawnApi class quite a lot, but I feel it's cleaner to keep it separate from CarPawn class, gives an easy-to-read interface`. 
Then, why `phys_*` in the name. Just name wise, it sounds appears to be something on the lines of `phys_vehicle_` PhysicsBody in multirotorpawnsimapi class, and hence the reader could misinterpret it.

rajat2004 replied:
I always did have a bit of a hard time in choosing variable names, and yes, the `phys_api` does seem to be misleading. Will rename it to `pawn_api_`

rajat2004 commented:
@madratman I had opened https://github.com/microsoft/AirSim/pull/2383 also which includes these commits so as to make it easier for other people to use the feature and get review on it
That one has the same commits as this, would you prefer to have the changes made in this PR or the other one? This one can be closed if merging the other PR will be okay for you, in that case will make the updates there itself directly

madratman commented:
Sure let's move to 2383. I'm fine with just a rename to pawn api wrt review. The other inconsistencies are wrt airsim codebase in general.

rajat2004 commented:
Ok great, then closing this PR and will update the other one
